Add plugin for scalafix

stuhood opened the pull request:
### Problem

[scalafix](https://scalacenter.github.io/scalafix/) supports a bunch of type-aware rewrites of scala code using [scalameta](http://scalameta.org/), but pants does not currently support it.

### Solution

Add support for `scalafix` in core, with support for semantic rewrites disabled by default. To enable semantic rewrites, it requires an scala compiler plugin, and so enabling it requires extra setup documented here in the `README.md` for scala.

### Result

Running something like:
```
./pants fmt.scalafix --semantic --rules=RemoveUnusedImports src/scala/org/pantsbuild/zinc/::
```
would strip unused imports and results in diffs like:
```
 import sbt.util.{
-  AbstractLogger,
+
   Level,
   Logger
 }
```

stuhood commented:
This will need a few more updates (namely a fix for https://github.com/scalacenter/scalafix/issues/176) before landing, so not adding reviewers yet.

stuhood commented:
This is now reviewable. Will add reviewers once I have green CI.

xeno-by commented on BUILD.tools:
```diff
@@ -48,3 +49,8 @@ jar_library(
+jar_library(
+  name = 'scalac-plugin-dep',
+  jars = [jar(org='org.scalameta', name='scalahost_{}'.format(SCALA_REV), rev='1.8.0')],
+)
```
Would it be possible to experiment with pre-release versions of scalameta without forking pants?

xeno-by replied:
After our recent release, this should now be `jars = [jar(org='org.scalameta', name='scalac-semanticdb_{}'.format(SCALA_REV), rev='2.0.0')],`.

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/BUILD:
```diff
@@ -0,0 +1,17 @@
+    'src/python/pants/goal:task_registrar',
+  ],
+  distribution_name='pantsbuild.pants.contrib.scalafix',
+  description='scalafix support for pants.',
```
@olafurpg What is the appropriate capitalization of "scalafix" here?

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,123 @@
+    args.append('--sourceroot={}'.format(results_dir))
+    args.append('--classpath={}'.format(':'.join(classpath)))
+    if self.get_options().config:
+      args.append('--config={}'.format(self.get_options().config))
```
@olafurpg What does take precedence: --config or --sourceroot/--classpath?

olafurpg replied:
--config does not have --classpath or --sourceroot, see
https://github.com/scalacenter/scalafix/blob/9b1b07021ba48a56906bfb26ec86c7cf209d32c7/scalafix-core/src/main/scala/scalafix/config/ScalafixConfig.scala#L14
and
https://github.com/scalacenter/scalafix/blob/9b1b07021ba48a56906bfb26ec86c7cf209d32c7/scalafix-cli/src/main/scala/scalafix/cli/ScalafixOptions.scala#L32

I admit it's a bit confusing to have different systems for cli options and the scalafix configuration.

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,123 @@
+  def _run(self, target, results_dir, classpath):
+    # We always operate on copies of the files, so we execute in place.
+    args = ['--in-place']
+    args.append('--sourceroot={}'.format(results_dir))
```
--sourceroot should point to the root of the build. Is this what `results_dir` stands for?

stuhood replied:
This is when we are _consuming_ a database, rather than when we're creating it. We're executing the rewrites on a files that have been cloned elsewhere (into a private `result_dir`),  rather that directly in the repo.

This allows for both ScalaFixCheck and ScalaFixFix to be implemented with the same codepaths.

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,123 @@
+    if self.get_options().rewrites:
+      args.append('--rewrites={}'.format(self.get_options().rewrites))
+    if self.get_options().level == 'debug':
+      args.append('--verbose')
```
Scalafix expects (or is supposed to expect) input files to be specified explicitly via the `-f` option. Is this option set anywhere in the current implementation?

xeno-by replied:
Most of scalafix rewrites need a semantic db to be generated for target files. Is it possible / does it make sense to provide sanity checks for that here?

olafurpg replied:
-f or --files is optional, unbound trailing arguments are interpreted as files `scalafix foo.scala`

olafurpg replied:
can you elaborate?

xeno-by replied:
This is somewhat ambiguous. What I meant to ask is whether it'd make sense to provide additional safeguards on the Pants side. Iirc, scalafix HEAD says "something something please recompile" if it doesn't find semantic dbs, but this message isn't going to help a Pants plugin user who tries to run scalafix without having a scalahost plugin in the first place.

stuhood replied:
I haven't actually seen the error, so not sure how good it is right now... but would love for the error from scalafix to be sufficiently helpful that I don't need to check for existence here.

benjyw replied:
Some sort of error message saying "make sure you set up the scalameta plugin using the blahblah BUILD.tools target or the blahblah option" would be helpful though. And that can't come from scalafix.

olafurpg replied:
scalafix will still in some cases silently fail on misconfigured semanticdbs, https://github.com/scalacenter/scalafix/issues/264 This is a bug, scalafix should report helpful error messages when things go wrong and report the appropriate exit code to accompany it. For example, scalafix will exit with code 32 when it encounters a "stale semanticdb" (file contents doesn't match semanticdb contents).

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,123 @@
+
+  def execute(self):
+    classpaths = self.context.products.get_data('runtime_classpath')
+    # NB: This task uses only the literal classpath of each target, so does not need
```
What's literal classpath?

stuhood replied:
Well... it's a thing that I don't have any other name for, frankly.

"transitive": All transitive dependencies of a compile unit.
"direct"/"string": Only the directly declared dependencies of a compile unit.
"literal"/???: Only the classfiles for the compile unit, and none of its dependencies.

If you have something else that you call that, happy to rename.

xeno-by replied:
Got it. I just did a bit of googling and didn't find a definition, so decided to ask you to better understand what's going on.

benjyw replied:
Something like "uses only the classes directly generated from the sources of the target"?

stuhood commented on BUILD.tools:
```diff
@@ -48,3 +49,8 @@ jar_library(
+jar_library(
+  name = 'scalac-plugin-dep',
+  jars = [jar(org='org.scalameta', name='scalahost_{}'.format(SCALA_REV), rev='1.8.0')],
+)
```
Yes. Tool definitions can be reconfigured on a repo by repo basis.

benjyw commented on contrib/scalafix/src/python/pants/contrib/scalafix/BUILD:
```diff
@@ -0,0 +1,17 @@
+contrib_plugin(
+  name='plugin',
+  dependencies=[
+    'contrib/node/src/python/pants/contrib/node:plugin',
```
?

stuhood replied:
Derp, thanks.

stuhood commented:
Hah. It looks like (some) of the CI failures are due to duels happening between scalafix and scalafmt, where scalafmt re-writes some files, and then scalafix rewrites them again from a cached copy which reverts the changes. Thinking about how best to address that (hopefully not by disabling caching for the `fmt.scalafix` task).

olafurpg commented:
@stuhood I've considered consolidating the two into a single cli tool. It should be possible to model scalafmt as a syntactic scalafix rewrite and add support to run scalafmt at the end. Do you think something like that might be useful?

olafurpg commented:
Heads up @stuhood, I've been polishing the cli lately to make it (hopefully) more intuitive. The main changes relate to how scalafix finds files. In a nutshell, scalafix will no longer infer the .scala files from the --classpath. Instead, you tell scalafix which files to fix and provide a --classpath to tell scalafix where to look for `.semanticdb` files.  This change makes `--include` unnecessary, since the file path args act now as --include for both syntactic and semantic rewrites.

olafurpg commented:
I'm hoping to release scalafix v0.5 around next week Tuesday. The v0.5 milestone has a few simple remaining issues https://github.com/scalacenter/scalafix/milestone/3 I have still yet to document the main changes, but they mostly are aimed to increase the quality-of-life for the cli so it might be worth the wait for this PR

stuhood commented:
We learned a bit more from b6cf3e5ae60756141709e199dbd6265b6b2c3417 about how formatters should work, so I've started down the path of implementing a shared base class for scalafmt and scalafix: now added here.

The `lint`/check versions of the tools need to run as a batch with xargs in order for the performance to not be atrocious (as scalafmt now does). Additionally, we need caching for the checks. But the `fmt`/rewrite versions of the tools can't use invalidation at all, for the reasons mentioned above: our fingerprinting doesn't react well to files changing out from under it.

The final interesting bit that I want to include in this change is that scalafix needs a classpath for some cases: it's become obvious internally that we cannot require a classpath for `lint, as it would make it way too slow to require compiling everything before running the checks. So we'll need to add a task-level option that enables or disables passing the classpath to scalafix (similar to 4b346175d71bbb08a15c7d8acf1d1640e58381a0): that will allow the task to run either with or without a dependency on the user classpath. That way you can install cheaper scalafix checks in `lint`, and more expensive checks elsewhere.

Unfortunately, I'm out until September 10th, so there won't be any movement from me until at least then.

olafurpg commented:
Scalafix 0.5.0-RC3 is out http://www.scala-lang.org/blog/2017/09/11/scalafix-v0.5.html, changelog https://github.com/scalacenter/scalafix/releases/tag/v0.5.0-RC3 I think the api (both library and cli flags) will remain stable for a while now, at least I don't see any breaking changes coming up anytime soon.

xeno-by commented on contrib/scalafix/README.md:
```diff
@@ -0,0 +1,54 @@
+Unlike `scalafmt` (which is included in the default distribution of pants), `scalafix` requires a
+compiler plugin, and is thus distributed as a contrib module with extra setup steps.
+
+The [scalameta](http://scalameta.org/) compiler plugin extracts semantic information at compile
```
`The [semanticdb](http://scalameta.org/) compiler plugin...`

xeno-by commented on contrib/scalafix/README.md:
````diff
@@ -0,0 +1,54 @@
+  name = 'scalac-plugin-dep',
+  jars = [jar(org='org.scalameta', name='scalahost_{}'.format(SCALA_REV), rev='1.8.0')],
+)
+```
````
See the suggested change to scalac-plugin-dep above.

xeno-by commented on contrib/scalafix/README.md:
```diff
@@ -0,0 +1,54 @@
+  ]
+
+scalac_plugins: [
+    'scalahost',
```
`    'semanticdb-scalac',`

xeno-by commented on contrib/scalafix/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Usage
+
+Tasks for both the `fmt` and `lint` goals are provided, each with the same set of options.
```
What do you think about calling these goals `fix` and `check` in accordance with how these actions are called in Scalafix? https://github.com/scalacenter/scalafix/blob/master/scalafix-core/shared/src/main/scala/scalafix/rule/Rule.scala.

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,129 @@
+  def register_options(cls, register):
+    super(ScalaFix, cls).register_options(register)
+    register('--skip', type=bool, fingerprint=True,
+             help='Skip running scalafix.')
```
Why do we need this option?

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,129 @@
+    classpaths = self.context.products.get_data('runtime_classpath')
+    # NB: While this task uses only the classes directly generated from the sources of each
+    # target, the content of the generated semantic DBs will be affected by changes to
+    # dependencies. Thus, we use invalidate_dependents=True.
```
Very good point!

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,129 @@
+                          invalidate_dependents=True) as invalidation_check:
+      for vt in invalidation_check.all_vts:
+        if not vt.valid:
+          self.context.log.debug('Fixing {}...'.format(vt.target.address.spec))
```
May I suggest changing the verb here depending on the goal that is being executed?

xeno-by commented on pants.ini:
```diff
@@ -151,17 +153,24 @@ args: [
   ]
 
+scalac_plugins: [
+    'scalahost',
```
`'semanticdb-scalac',`

xeno-by commented on src/python/pants/backend/jvm/tasks/scalafmt.py:
```diff
@@ -115,6 +124,51 @@ def calculate_sources(self, targets):
+                          JarDependency(org='com.geirsson',
+                                        name='scalafmt-cli_2.11',
+                                        rev='1.0.0-RC4')
+                          ])
```
Looks like an outdated version.

xeno-by commented on contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py:
```diff
@@ -0,0 +1,129 @@
+      shutil.copy(src, dst)
+
+    # Execute.
+    result = self.runjava(classpath=self.tool_classpath('scalafix'),
```
Am I right in thinking that `scalafix` is supposed to be bundled with Pants?

stuhood commented:
Alright, I've made a bunch of changes:
1) scalafmt and scalafix now extend a `ScalaRewriteBase` class to provide the generic harnessy bits
2) added a `--transitive` option on `ScalaRewriteBase`
3) added a `--semantic` flag for scalafix, to toggle whether the user classpath (and thus semanticdbs) are required.

I'm going to apply @xeno-by's feedback with regard to versioning. I also think that I can remove the contrib module and move this into core by changing the default of the `--semantic` flag to `False`: that would allow the tasks to be installed by default in a mode that doesn't require the compiler plugins (or compilation at all)... parties interested in enabling semantic rewrites would still need to install the `semanticdb-scalac` plugin as documented.

stuhood commented:
Test flakiness. Merging.
